src/ir/effects.h

```diff
@@ -466,43 +481,61 @@ class EffectAnalyzer {
         return;
       }
 
+      const EffectAnalyzer* targetEffects = nullptr;
+      if (parent.funcEffectsMap) {
+        auto iter = parent.funcEffectsMap->find(curr->target);
+        if (iter != parent.funcEffectsMap->end()) {
+          targetEffects = &iter->second;
+        }
+      }
+
       if (curr->isReturn) {
         parent.branchesOut = true;
+        // When EH is enabled, any call can throw.
+        if (parent.features.hasExceptionHandling() &&
+            (!targetEffects || targetEffects->throws())) {
+          parent.hasReturnCallThrow = true;
+        }
       }
 
-      if (parent.funcEffectsMap) {
-        auto iter = parent.funcEffectsMap->find(curr->target);
-        if (iter != parent.funcEffectsMap->end()) {
-          // We have effect information for this call target, and can just use
-          // that. The one change we may want to make is to remove throws_, if
-          // the target function throws and we know that will be caught anyhow,
-          // the same as the code below for the general path.
-          const auto& targetEffects = iter->second;
-          if (targetEffects.throws_ && parent.tryDepth > 0) {
-            auto filteredEffects = targetEffects;
-            filteredEffects.throws_ = false;
-            parent.mergeIn(filteredEffects);
-          } else {
-            // Just merge in all the effects.
-            parent.mergeIn(targetEffects);
-          }
-          return;
+      if (targetEffects) {
+        // We have effect information for this call target, and can just use
+        // that. The one change we may want to make is to remove throws_, if the
+        // target function throws and we know that will be caught anyhow, the
+        // same as the code below for the general path. We can always filter out
+        // throws for return calls because they are already more precisely
+        // captured by `hasReturnCallThrow`.
+        if (targetEffects->throws_ && (parent.tryDepth > 0 || curr->isReturn)) {
```

Comment:
The `isReturn` part here is not clear to me, because `hasReturnCallThrow` only takes effect in `walk(Function*)`. But most users of this class do not use that, that is, they do not walk an entire function. For them, this would be removing the throws effect here - isn't that dangerous?

Concretely, Vacuum will remove this dropped call, I worry:
```wat
(func $caller
  (return_call $throw)
)
(func $throw
  (throw ..)
)
```
I worry that is the case because Vacuum computes effects for each expression it sees, including that `return_call`. The `return_call` happens to be the entire function body, but Vacuum is just walking all the expressions and doesn't know that, so it calls `walk(Expression*)` and not `walk(Function*)`.

That is, I think it is unsafe to leave things here for function-level handling because we are not always looking at the entire function.

Reply:
I hope that the `branchesOut` effect will prevent that kind of problem. Do you think it will?

Reply:
I suppose `branchesOut` fixes that specific case, good point. But if we had a testcase that was sensitive to `throws` specifically we could be in trouble. I admit I can't think of one, but I think the principle remains that we should not dpeendepend on `hasReturnCallThrow` to save us because that only runs at the function level by design, and we are not always scanning at that level.

Reply:
Yes, I think that's right @vouillon - after thinking more on this I agree. Overall I'd explain the rationale like this:

1. `return_call` is basically like syntax sugar for a branch to an outer block at the function's exit where a call occurs. This is the key thing to keep in mind.
2. The call will definitely execute, so it's fine to just apply its effects here, normally.
3. However, a noticeable difference is when the call's effects can be removed due to scoping, which atm includes only try-catch (but may include more in future specs). In that case we do need to model the effect as not occurring here but rather as occurring in the outer block.
4. To do that, we can remove `throws_` and keep the normal `branchesOut`, which exactly represents a branch to another place, and that place is the outer scope that has the `throws_`. Users of the code that just consider a single expression will call `walk(Expression*)` and they cannot remove `branchesOut` effects because it is a branch to some place that can have arbitrary effects (even more powerful than `throws_`, and which they don't see), so this is safe.
5. And, in `walk(Function*)` we apply that outer block and the effects of the calls that occur there, that is, users of the code that want to consider entire functions must use that form. It would be a bug for them not to, so that is something users must be careful about.

So I think this PR is correct as it is but perhaps @tlively we can add that rationale or something like it?

Reply:
Most of this was already captured in the comments, but I just pushed a commit adding extra detail. Let me know if you have suggestions for further clarifications.